### PR TITLE
Refactoring of interfaces and multi type branching from another column.

### PR DESCRIPTION
**Pull request by mbaric758**
- Introduces type branching based on value from another column  #121 
- Fixes bugs for usage of multi type transformers on subjects #127 
- Improves existing tests with edge cases and adds tests for new features
- Updates docs.
- Adds new interfaces for defining logic of cell value extraction and type definition #123  
- Introduces more synonyms to be used in the YAML mapping #117 
- Decouples the parsing stage into three distinct steps.

### CODE REVIEW

File: docs/README.rst [hunk 1]
```diff
@@ -596,6 +596,72 @@ subject type ``line`` with different edge types. The cell values
...
+                    to_object: person
+                    via_relation: will_not_sit
+
+With this mapping, we want to map the column ``furniture`` to the node types ``kitchen_furniture`` and
+``rest_of_house_furniture`` based on their localisation. The localisation of each piece of furniture is extracted from
+the column ``localisation``. The mapping uses the ``match`` clause to apply different type mappings based on the
+localisation of the furniture, similarly as it was done in the previous example. This time, however, the ``match`` clause
+needs to look at the values of another column - ``localisation``, to determine the type of the node to be created.
+In this case, we use the keyword ``match_type_from_column`` to indicate that the type of the node to be created depends
+on the value of the ``localisation`` column. The ``id_from_column`` keyword indicates that the id of the node to be
```
Comment (jdreo) [inline docs/README.rst:648]:
```suggestion
on the value of the ``localisation`` column. The ``id_from_column`` keyword indicates that the identifier of the node to be
```

File: docs/README.rst [hunk 2]
```diff
@@ -596,6 +596,72 @@ subject type ``line`` with different edge types. The cell values
...
+created should be taken from the column ``furniture``.
+
+Next, we want to map the column ``name`` to the node type ``person``, and define the edge type based on whether the
+furniture will be sat on or not. We extract the name of the person from the column ``name``, using the ``id_form_column``
```
Comment (jdreo) [inline docs/README.rst:652]:
```suggestion
furniture will be "sat on" or not. We extract the name of the person from the column ``name``, using the ``id_form_column``
```

File: docs/README.rst [hunk 3]
```diff
@@ -714,6 +781,10 @@ of several node types:
                - case
                - phenotype
 
+Note that the properties declared in the BioCypher ``schema_config.yaml`` must match the properties declared in the mapping configuration file.
+Furthermore, when declaring the properties in the schema configuration file, take care that the property must always be a
+string (``str``) type - in order to avoid errors when importing the data into the Neo4j graph database.
```
Comment (jdreo) [inline docs/README.rst:786]:
```suggestion
string (``str``) type, which is the only supported property type so far.
```

File: docs/readme_sections/how_to.rst [hunk 4]
```diff
@@ -38,6 +38,10 @@ of several node types:
                - case
                - phenotype
 
+Note that the properties declared in the BioCypher ``schema_config.yaml`` must match the properties declared in the mapping configuration file.
+Furthermore, when declaring the properties in the schema configuration file, take care that the property must always be a
+string (``str``) type - in order to avoid errors when importing the data into the Neo4j graph database.
```
Comment (jdreo) [inline docs/readme_sections/how_to.rst:43]: idem

File: docs/readme_sections/mapping_api.rst [hunk 5]
```diff
@@ -422,6 +422,72 @@ subject type ``line`` with different edge types. The cell values
 values ``sensitivity`` and ``productivity`` would be mapped to the node
 type ``noun`` via the edge type ``line_is_noun``.
 
+Type branching based on value from another column
```
Comment (jdreo) [inline docs/readme_sections/mapping_api.rst:425]: We should not have duplicated documentation. If it's in the doc, you can safely discard what's in the README above.

File: src/ontoweaver/make_value.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+        expected_att = self.__dict__
+        self.__dict__.update(**kwargs)
+
+        expected_keys = {k: v for k, v in expected_att.items() if not k.startswith('__')}
+        for k, v in expected_keys.items():
+            if v is None:
+                self.error(f"Attribute {k} not set.", section=f"{self.__class__.__name__}.check", exception=exceptions.TransformerDataError)
+
+    @abc.abstractmethod
+    def select(self, columns, row, i):
```
Comment (jdreo) [inline src/ontoweaver/make_value.py:36]: Let's follow the same naming convention than for the LabelMaker class:
```suggestion
    def call(self, columns, row, i):
```

File: src/ontoweaver/transformer.py [hunk 7]
```diff
@@ -46,22 +47,38 @@ def __call__(self, row, i):
...
+            super().__init__(raise_errors)
+
+        def select(self, columns, row, i):
+
+            for key in columns:
+                items = str(row[key]).split(self.separator)
+                for item in items:
+                    yield item
+
+    def __init__(self, properties_of, value_maker = ValueMaker(), label_maker = None, branching_properties = None, columns=None, output_validator: validate.OutputValidator = None, raise_errors = True, **kwargs):
```
Comment (jdreo) [inline src/ontoweaver/transformer.py:66]: We should have an explicit declaration of the default arguments in the constructor's signature…
```suggestion
    def __init__(self, properties_of, value_maker = ValueMaker(), label_maker = None, branching_properties = None, columns=None, output_validator: validate.OutputValidator = None, raise_errors = True, separator = ",", **kwargs):
```

File: src/ontoweaver/transformer.py [hunk 8]
```diff
@@ -46,22 +47,38 @@ def __call__(self, row, i):
...
+            label_maker: the LabelMaker object used for handling the creation of the output of the transformer. Default is None.
             branching_properties: in case of branching on cell values, the dictionary holding the properties for each branch.
             columns: The columns to be processed.
             sep: Character(s) to use for splitting.
             output_validator: the OutputValidator object used for validating transformer output.
             raise_errors: if True, the caller is asking for raising exceptions when an error occurs
         """
-        super().__init__(properties_of, branching_properties, columns, output_validator, raise_errors = raise_errors, **kwargs)
+        super().__init__(properties_of, value_maker, label_maker, branching_properties, columns, output_validator,
+                         raise_errors=raise_errors, **kwargs)
```
Comment (jdreo) [inline src/ontoweaver/transformer.py:81]: … followed by an explicit initialization of the corresponding member variable…
```suggestion
                         raise_errors=raise_errors, **kwargs)
                         if not separator: # Neither empty string or None
                             self.error(f"The `separator` paramater of the `{self.__name__}` transformer cannot be an empty string.")
                         self.separator = separator
```

File: src/ontoweaver/make_labels.py [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+        self.edge_type = edge_type
+        self.target_node_type = target_node_type
+        self.target_element_properties = properties_of
+
+class LabelMaker(errormanager.ErrorManager, metaclass=abc.ABCMeta):
+    """Interface for defining the correct labels (types) for each extracted value.
+    The class is responsible for validating the extracted cell value and returning the appropriate label.
+    """
+
+    def __init__(self, raise_errors: bool = True):
```
Comment (jdreo) [inline src/ontoweaver/make_labels.py:35]: This should get `**kwargs` instead of `call`.

File: src/ontoweaver/make_labels.py [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+        """
+        Initializes the LabelMaker object.
+
+        Args:
+            raise_errors (bool): if True, will raise an exception when an error is encountered, else, will log the error and try to proceed.
+        """
+        super().__init__(raise_errors)
+
+    def check_attributes(self, **kwargs):
```
Comment (jdreo) [inline src/ontoweaver/make_labels.py:44]: This function may not be necessary if `kwargs` goes through `__init__` and Python handle member variables initialization checks.

File: src/ontoweaver/make_labels.py [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+        if missing_keys:
+            for k in missing_keys:
+                self.error(
+                    f"Attribute {k} not set.",
+                    section=f"{self.__class__.__name__}.check",
+                    exception=exceptions.TransformerDataError
+                )
+
+    @abc.abstractmethod
+    def call(self, validate, returned_value, multi_type_dict, branching_properties = None, **kwargs):
```
Comment (jdreo) [inline src/ontoweaver/make_labels.py:70]: `kwargs` may not be necessary here, as the call is responsible for handling the data, and the constructor for handling the parameters.

File: src/ontoweaver/make_value.py [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+from . import errormanager
+from . import exceptions
+
+logger = logging.getLogger("ontoweaver")
+class ValueMaker(errormanager.ErrorManager, metaclass=abc.ABCMeta):
+    """
+    Interface for value makers that are used to define the extraction logic for each of the transformer classes.
+    """
+    def __init__(self, raise_errors: bool = True):
```
Comment (jdreo) [inline src/ontoweaver/make_value.py:12]: Same rationale about `kwargs` than for `LabelMaker`: the constructor should take `kwargs`, not the call function, and there is no need for checking attributes.

File: src/ontoweaver/make_value.py (hunk 6 above)
Comment (jdreo) [reply src/ontoweaver/make_value.py:36]: Or maybe even use the `__call__` interface, since we should not need `check_attributes`.

File: src/ontoweaver/tabular.py [hunk 13]
```diff
@@ -644,6 +645,7 @@ def make_transformer_class(self, transformer_type, multi_type_dictionary = None,
                             output_validator=output_validator,
                             multi_type_dict = multi_type_dictionary,
                             branching_properties = branching_properties,
+                            label_maker = label_maker,
                             raise_errors = raise_errors,
                             **kwargs)
```
Comment (jdreo) [inline src/ontoweaver/tabular.py:650]: `kwargs` is already passed to the constructor, this is what we want indeed.

File: src/ontoweaver/transformer.py [hunk 14]
```diff
@@ -74,29 +91,41 @@ def __call__(self, row, i):
         Yields:
             str: Each split item from the cell value.
         """
-        for key in self.columns:
-            items = str(row[key]).split(self.separator)
-            for item in items:
-                yield self.create(item)
+        for value in self.value_maker(self.columns, row, i, separator = self.separator):
```
Comment (jdreo) [inline src/ontoweaver/transformer.py:94]: … or else `self.separator` is not defined here.

File: src/ontoweaver/transformer.py [hunk 15]
```diff
@@ -135,7 +169,7 @@ def __init__(self, properties_of, branching_properties = None, columns=None, out
             raise_errors: if True, the caller is asking for raising exceptions when an error occurs
         """
         super().__init__(properties_of, branching_properties, columns, output_validator, multi_type_dict,
-                         raise_errors = raise_errors,**kwargs)
+                         raise_errors=raise_errors, **kwargs)
```
Comment (jdreo) [inline src/ontoweaver/transformer.py:172]: Missing an explicit `self.format_string` initialization.

File: src/ontoweaver/transformer.py [hunk 16]
```diff
@@ -232,23 +287,35 @@ def __call__(self, row, i):
...
 
 class translate(base.Transformer):
     """Translate the targeted cell value using a tabular mapping and yield a node with using the translated ID."""
 
-    def __init__(self, properties_of, branching_properties = None, columns=None, output_validator: validate.OutputValidator = None, multi_type_dict = None, raise_errors = True, **kwargs):
+    class ValueMaker(make_value.ValueMaker):
+        def __init__(self, raise_errors: bool = True):
+            super().__init__(raise_errors)
+
+        def select(self, columns, row, i, **kwargs):
```
Comment (jdreo) [inline src/ontoweaver/transformer.py:302]: `kwargs` should rather go into the constructor.

File: src/ontoweaver/transformer.py [hunk 17]
```diff
@@ -260,9 +327,9 @@ def __init__(self, properties_of, branching_properties = None, columns=None, out
             raise_errors: if True, the caller is asking for raising exceptions when an error occurs
             kwargs: Additional arguments to pass to Pandas' read_csv (if "sep=TAB", reads the translations_file as tab-separated).
         """
-        super().__init__(properties_of, branching_properties, columns, output_validator, multi_type_dict,
-                         raise_errors = raise_errors, **kwargs)
-        self.map = map(properties_of, branching_properties, columns, output_validator, multi_type_dict, **kwargs)
+        super().__init__(properties_of, value_maker, label_maker, branching_properties, columns, output_validator,
+                         multi_type_dict, raise_errors=raise_errors, **kwargs)
+        self.map = map(properties_of, value_maker, label_maker, branching_properties, columns, output_validator, multi_type_dict, **kwargs)
```
Comment (jdreo) [inline src/ontoweaver/transformer.py:332]: I don't understand what `map` is/does.

File: tests/test_multi_type_transformer.py [hunk 18]
```diff
@@ -0,0 +1,43 @@
+# # FIXME: The following code is for the test of the multi type transformers. Due to it causing testing errors it is commented out and to be fixed.
```
Comment (jdreo) [inline tests/test_multi_type_transformer.py:1]: :fearful:

**Review by jdreo — CHANGES_REQUESTED** [review]
Huge work!

Since we talked about getting rid of the `check_attributes` functions by clearly separating parameter initialization in constructors and data in call interfaces, I wrote my comments that way. If you found that this was not possible, then maybe you should add comments indicating why in the root classes.

File: tests/test_multi_type_transformer.py (hunk 18 above)
Comment (mbaric758) [reply tests/test_multi_type_transformer.py:1]: Yes... I still have that error and no idea why. The test itself works and is correct, but causes some of the other tests to fail. 😩

File: src/ontoweaver/transformer.py (hunk 17 above)
Comment (mbaric758) [reply src/ontoweaver/transformer.py:332]: The translate transformer uses `map` after translating the labels to do the actual extraction of data.

File: src/ontoweaver/make_value.py (hunk 6 above)
Comment (jdreo) [reply src/ontoweaver/make_value.py]: What kind of errors?

Comment (mbaric758) [reply src/ontoweaver/make_value.py]: I was wrong - the way it should have been done is to instantiate the class once all the parameters are already passed to the `__init__`. This is how it is now, with no use of the parameter checking functions.

**Review by jdreo — APPROVED** [review]
Let's do it!
